src/librbd/Journal.cc

```diff
@@ -822,51 +822,63 @@ uint64_t Journal<I>::append_io_event(journal::EventEntry &&event_entry,
+  return append_io_events(event_entry.get_event_type(), bufferlists, requests, offset,
                           length, flush_entry);
 }
 
 template <typename I>
 uint64_t Journal<I>::append_io_events(journal::EventType event_type,
-                                      const Bufferlists &bufferlists,
+                                      Bufferlists *bufferlists,
                                       const AioObjectRequests &requests,
                                       uint64_t offset, size_t length,
                                       bool flush_entry) {
-  assert(!bufferlists.empty());
+  assert(!bufferlists->empty());
 
-  Futures futures;
   uint64_t tid;
   {
     Mutex::Locker locker(m_lock);
     assert(m_state == STATE_READY);
 
-    Mutex::Locker event_locker(m_event_lock);
     tid = ++m_event_tid;
     assert(tid != 0);
 
-    for (auto &bl : bufferlists) {
-      assert(bl.length() <= m_max_append_size);
-      futures.push_back(m_journaler->append(m_tag_tid, bl));
-    }
-    m_events[tid] = Event(futures, requests, offset, length);
-  }
-
-  CephContext *cct = m_image_ctx.cct;
-  ldout(cct, 20) << this << " " << __func__ << ": "
-                 << "event=" << event_type << ", "
-                 << "new_reqs=" << requests.size() << ", "
-                 << "offset=" << offset << ", "
-                 << "length=" << length << ", "
-                 << "flush=" << flush_entry << ", tid=" << tid << dendl;
+    FunctionContext *ctx = new FunctionContext(
+        [this, bufferlists, requests, offset, length, tid, event_type,
+         flush_entry](int r) {
+          Futures futures;
+          for (auto &bl : *bufferlists) {
+            assert(bl.length() <= m_max_append_size);
+            futures.push_back(m_journaler->append(m_tag_tid, bl));
+          }
+          delete bufferlists;
+
+          Mutex::Locker event_locker(m_event_lock);
+          m_events[tid] = Event(futures, requests, offset, length);
+          m_events_cond.Signal();
+
+          CephContext *cct = m_image_ctx.cct;
+          ldout(cct, 20) << this << " " << __func__ << ": "
+                         << "event=" << event_type << ", "
+                         << "new_reqs=" << requests.size() << ", "
+                         << "offset=" << offset << ", "
+                         << "length=" << length << ", "
+                         << "flush=" << flush_entry << ", tid=" << tid << dendl;
+
+          Context *on_safe = create_async_context_callback(
+            m_image_ctx, new C_IOEventSafe(this, tid));
+          if (flush_entry) {
+            futures.back().flush(on_safe);
+          } else {
+            futures.back().wait(on_safe);
+          }
+        }
+    );
 
-  Context *on_safe = create_async_context_callback(
-    m_image_ctx, new C_IOEventSafe(this, tid));
-  if (flush_entry) {
-    futures.back().flush(on_safe);
-  } else {
-    futures.back().wait(on_safe);
+    m_work_queue->queue(ctx);
```

Comment:
I am curious as to how this helps to reduce latency of IO completions. A ThreadPool thread context will be used to start processing an IO operation, this will enqueue a new op for later processing, and then the current context will finish the AioCompletion once the IO hits the in-memory cache.  However, now a thread context will get utilized to append the journal event -- which is now one less thread available for IO processing.  Did you increase your thread pool to 2 times the IO queue depth?